File: windows/SeedSeeker.Tests/ItemMappingsTests.cs
using System.Text;
using Xunit;

namespace SeedSeeker.Tests;

public sealed class ItemMappingsTests
{
    [Fact]
    public void NativeScoutReturnsEveryMappingInGameClassOrder()
    {
        var artwork = ItemMappingArtwork.Shared;
        Assert.Equal(new[] { "potions", "rings", "scrolls" }, artwork.Categories.Keys.Order());
        Assert.All(artwork.Categories.Values, art => Assert.Equal(12, art.IconSizes.Length));
        var engine = new NativeEngine();
        var world = engine.Scout("ABC-DEF-GHI", 0);
        var mappings = Assert.IsType<ScoutItemMappings>(world.ItemMappings);
        Assert.Equal(new ScoutItemMapping("Scroll of upgrade", "TIWAZ", 315), mappings.Scrolls[0]);
        foreach (var group in new[] { mappings.Scrolls, mappings.Potions, mappings.Rings })
        {
            Assert.Equal(12, group.Count);
            Assert.Equal(12, group.Select(x => x.SpriteIndex).Distinct().Count());
            Assert.Equal(12, group.Select(x => x.Appearance).Distinct().Count());
        }
        Assert.Equal(world.Gems.Ordinals.Select(x => (int)x), mappings.Rings.Select(x => x.SpriteIndex - 224));
        var challenged = engine.Scout(world.Seed, 1, trinket: "none").ItemMappings!;
        Assert.Equal(mappings.Scrolls, challenged.Scrolls);
        Assert.Equal(mappings.Potions, challenged.Potions);
        Assert.Equal(mappings.Rings, challenged.Rings);
    }

    [Fact]
    public void RoomSummariesDecodeWithoutChangingLegacyPackets()
    {
        var prefix = Packet("SSC8");
        var tail = new Writer(); tail.U8(2); tail.U8(7); tail.U8(1); tail.Text("garden");
        tail.U8(17); tail.U8(1); tail.Text("secret_garden");
        var world = NativeEngine.DecodeScout([.. prefix, .. tail.Finish()]);
        Assert.Equal(new[] { "garden" }, world.FloorRooms![7]);
        Assert.Equal(new[] { "secret_garden" }, world.FloorRooms[17]);
        Assert.Empty(NativeEngine.DecodeScout(Packet()).FloorRooms!);
        Assert.False(world.IsFarmingFloor(7));
        foreach (byte[] invalid in new byte[][] { [], [21], [1, 0, 0], [1, 5, 0], [1, 25, 0],
            [2, 7, 0, 7, 0], [2, 17, 0, 7, 0], [1, 7, 1], [1, 7, 1, 0, 0], [0, 0] })
            Assert.Throws<InvalidDataException>(() => NativeEngine.DecodeScout([.. prefix, .. invalid]));
        Assert.Throws<InvalidDataException>(() => NativeEngine.DecodeScout([.. prefix, .. tail.Finish()[..^1]]));
    }

    private static byte[] Packet(string version = "SSC7", string mutation = "")
    {
        var w = new Writer();
        w.Bytes(Encoding.UTF8.GetBytes(version)); w.U8(11); w.Bytes(Encoding.UTF8.GetBytes("AAA-AAA-AAA"));
        w.Bytes(Enumerable.Range(0, 12).Select(x => (byte)x)); w.U8(0); w.U16(0);
        if (version == "SSC3") return w.Finish();
        w.U8(17); foreach (var item in ItemCatalog.For(ItemKind.Trinket)) w.Text(item.Id);
        if (version == "SSC4") return w.Finish();
        w.U8(0);
        if (version == "SSC5") return w.Finish();
        w.Text("");
        if (version == "SSC6") return w.Finish();
        foreach (var spriteBase in new[] { 304, 352, 224 })
        for (var index = 0; index < 12; index++)
        {
            w.Text(mutation == "blank" ? "" : $"Item {index}");
            w.Text(mutation == "duplicate" ? "Same" : $"Appearance {index}");
            w.U16(mutation == "range" ? 65535 : spriteBase + (mutation == "ring" && spriteBase == 224 ? (index + 1) % 12 : index));
        }
        return w.Finish();
    }

    [Fact]
    public void LegacyPacketsOmitMappingsAndMalformedExtensionsFail()
    {
        foreach (var version in new[] { "SSC3", "SSC4", "SSC5", "SSC6" })
            Assert.Null(NativeEngine.DecodeScout(Packet(version)).ItemMappings);
        var packet = Packet();
        Assert.NotNull(NativeEngine.DecodeScout(packet).ItemMappings);
        Assert.Throws<InvalidDataException>(() => NativeEngine.DecodeScout(packet[..^1]));
        Assert.Throws<InvalidDataException>(() => NativeEngine.DecodeScout([.. packet, 0]));
        foreach (var mutation in new[] { "blank", "duplicate", "range", "ring" })
            Assert.Throws<InvalidDataException>(() => NativeEngine.DecodeScout(Packet(mutation: mutation)));
    }
}

File: macos/SeedSeeker/Tests/SeedSeekerKitTests/ItemMappingsTests.swift
import Foundation
import XCTest
@testable import SeedSeekerKit

final class ItemMappingsTests: XCTestCase {
    func testNativeScoutReturnsAllMappingsIndependentOfChallengesAndTrinkets() async throws {
        let artwork = ItemMappingArtwork.shared
        XCTAssertEqual(artwork.categories.keys.sorted(), ["potions", "rings", "scrolls"])
        XCTAssertTrue(artwork.categories.values.allSatisfy { $0.iconSizes.count == 12 })
        let engine = ProductionSeedFinderEngine()
        let world = try await engine.scoutSeed("ABC-DEF-GHI", challenges: 0)
        let mappings = try XCTUnwrap(world.itemMappings)
        XCTAssertEqual(mappings.scrolls[0], ScoutItemMapping(name: "Scroll of upgrade", appearance: "TIWAZ", spriteIndex: 315))
        for group in [mappings.scrolls, mappings.potions, mappings.rings] {
            XCTAssertEqual(group.count, 12)
            XCTAssertEqual(Set(group.map(\.spriteIndex)).count, 12)
            XCTAssertEqual(Set(group.map(\.appearance)).count, 12)
        }
        XCTAssertEqual(mappings.rings.map { $0.spriteIndex - 224 }, world.ringGems.ordinals)
        let changed = try await engine.scoutSeed(world.seed, challenges: 1, query: nil, trinket: "none")
        XCTAssertEqual(mappings, changed.itemMappings)
    }

    func testRoomSummariesDecodeWithoutChangingLegacyPackets() throws {
        let prefix = packet(version: "SSC8")
        let tail = Data([2, 7, 1, 0, 6] + Array("garden".utf8) + [17, 1, 0, 13] + Array("secret_garden".utf8))
        let world = try ScoutCodec.decode(prefix + tail)
        XCTAssertEqual(world.floorRooms, [7: ["garden"], 17: ["secret_garden"]])
        XCTAssertTrue(try ScoutCodec.decode(packet()).floorRooms.isEmpty)
        XCTAssertFalse(world.isFarmingFloor(7)) // Room data alone does not imply Dark.
        for tail: [UInt8] in [[], [21], [1, 0, 0], [1, 5, 0], [1, 25, 0],
                              [2, 7, 0, 7, 0], [2, 17, 0, 7, 0], [1, 7, 1], [1, 7, 1, 0, 0], [0, 0]] {
            XCTAssertThrowsError(try ScoutCodec.decode(prefix + Data(tail)))
        }
        XCTAssertThrowsError(try ScoutCodec.decode((prefix + tail).dropLast()))
    }

    private func packet(version: String = "SSC7", mutation: String = "") -> Data {
        var data = Data()
        func bytes<S: Sequence>(_ value: S) where S.Element == UInt8 { data.append(contentsOf: value) }
        func u8(_ value: Int) { data.append(UInt8(value)) }
        func u16(_ value: Int) { bytes([UInt8(value >> 8), UInt8(value & 255)]) }
        bytes(version.utf8); u8(11); bytes("AAA-AAA-AAA".utf8)
        bytes((0..<12).map(UInt8.init)); u8(0); u16(0)
        if version == "SSC3" { return data }
        u8(17)
        for item in ItemCatalog.trinkets { u16(item.id.utf8.count); bytes(item.id.utf8) }
        if version == "SSC4" { return data }
        u8(0)
        if version == "SSC5" { return data }
        u16(0)
        if version == "SSC6" { return data }
        for base in [304, 352, 224] {
            for index in 0..<12 {
                let name = mutation == "blank" ? "" : "Item \(index)"
                let appearance = mutation == "duplicate" ? "Same" : "Appearance \(index)"
                u16(name.utf8.count); bytes(name.utf8)
                u16(appearance.utf8.count); bytes(appearance.utf8)
                u16(mutation == "range" ? 65535 : base + (mutation == "ring" && base == 224 ? (index + 1) % 12 : index))
            }
        }
        return data
    }

    func testLegacyPacketsOmitMappingsAndMalformedExtensionsFail() throws {
        for version in ["SSC3", "SSC4", "SSC5", "SSC6"] {
            XCTAssertNil(try ScoutCodec.decode(packet(version: version)).itemMappings)
        }
        let valid = packet()
        XCTAssertNotNil(try ScoutCodec.decode(valid).itemMappings)
        XCTAssertThrowsError(try ScoutCodec.decode(valid.dropLast()))
        XCTAssertThrowsError(try ScoutCodec.decode(valid + Data([0])))
        for mutation in ["blank", "duplicate", "range", "ring"] {
            XCTAssertThrowsError(try ScoutCodec.decode(packet(mutation: mutation)))
        }
    }
}
